### PR TITLE
Change implementation and test to pass in the user's pubkey that call…

### PR DESCRIPTION
**Pull request by psatyajeet**
# Description

In the `add_gif` method, we set `user_address` on the newly created `ItemStruct` (which is meant to store a link to the GIF and the address of the user who submitted the GIF) to the public key of the `baseAddress`. My understanding is that this `baseAddress` is how Solana looks up the Account that stores state for the application. We want to change implementation and test to use the user's pubkey that called `add_gif` to the `AddGif` `Context` so that we can save that to the `ItemStruct.user_address`
## Changes

* Solana_And_Web3/en/Section_2/Lesson_4_Store_Structs_On_Program.md
  *   Update `pub fn add_gif`to reference the user from `ctx: Context<AddGif>` and change the value for `user_address` in `ItemStruct` to `*user.to_account_info().key` instead of `*base_account.to_account_info().key`
  * Update the test call `await program.rpc.addGif(...` by passing in `user: provider.wallet.publicKey,` in addition to `baseAddress`

### CODE REVIEW

**Comment by adilanchian** [comment]
Hey @psatyajeet - 

Thanks so much for the PR here :). So, I'm going to drop some thoughts, but will let @farzaa add onto this as well.

Solana uses accounts to store data per program - you are indeed right. In the normal instance, you are creating a new account to interact with a program. For example, if you hit the GIF Portal multiple times you would use your user account (as you also stated!)

The problem here is your data is siloed in your account which means you would never be able to see any of the other GIFs that others submitted. That is why in this project we have a `baseAccount` that all users interact with. All the data is stored in that account 

TL;DR - your change here is not incorrect, but for the functionality we are going for in this project, this wouldn't be ideal! Let me know if you have any thoughts or concerns here :)

**Comment by psatyajeet** [comment]
Appreciate the comment @adilanchian ! 

Forgive me if I'm missing something, but I believe this change doesn't affect the `baseAccount` that all users interact with. The `gifList` would still be stored in the `baseAccount` so all users would be pulling the `gifList` from a single `baseAccount`. 

The only change being made here is that the `userAddress` in `ItemStruct` (`gifList` is `Vec<ItemStruct>`) should be set to the user that submitted the GIF rather than the `baseAddress`. 

The problem with the existing functionality is that every single GIF is being associated with the`baseAccount`. For example, if we wanted to add a tip button on every GIF to allow tipping the user that submitted the GIF, the `userAddress` on every entry would be the address of the `baseAccount` so we'd be tipping the `baseAccount` every single time.

**Comment by ItsJamesMurray** [comment]
@psatyajeet this was also very confusing to me as well. I am wanting to show the SOL address of the user that submitted a GIF, but  running `item.userAddress.toString()` as suggested in Section 4 Lesson 1 it is showing the baseAccount for every single GIF regardless of whoever submits the GIF. 

Did you make your proposed change on your end successfully?

**Comment by psatyajeet** [comment]
> @psatyajeet this was also very confusing to me as well. I am wanting to show the SOL address of the user that submitted a GIF, but running `item.userAddress.toString()` as suggested in Section 4 Lesson 1 it is showing the baseAccount for every single GIF regardless of whoever submits the GIF.
> 
> Did you make your proposed change on your end successfully?

I did. Yeah, the issue you're pointing out is the one I was having too. my solana program is here: https://github.com/psatyajeet/buildspace-solana-myepicproject and the React project is here: https://github.com/psatyajeet/sol-footy-gifs

**Comment by adilanchian** [comment]
Thanks for the heads up @psatyajeet !

Ya this makes more sense now. I chatted about it with @farzaa and he is going to do some testing on his end before we merge this in!

Thanks again for the awesome PR :)

**Comment by farzaa** [comment]
@psatyajeet @ItsJamesMurray

This looks correct. Running the updated code on my machine now to make sure it doesn't break anything

**Comment by farzaa** [comment]
10/10 find @psatyajeet. Confirmed on my machine. DM me on Discord and I'll send you $100 in ETH for this epic find!

**Comment by farzaa** [comment]
@all-contributors please add @psatyajeet for doc
